opentelemetry-api/src/opentelemetry/trace/__init__.py

```diff
@@ -410,23 +411,99 @@ def use_span(
+    """
+
+    def __init__(
+        self, get_current_tracer: typing.Callable[[], "Tracer"],
+    ) -> None:
+        self._get_current_tracer = get_current_tracer
+
+    def start_span(
+        self,
+        name: str,
+        context: typing.Optional[Context] = None,
+        kind: SpanKind = SpanKind.INTERNAL,
+        attributes: types.Attributes = None,
+        links: typing.Sequence[Link] = (),
+        start_time: typing.Optional[int] = None,
+        record_exception: bool = True,
+        set_status_on_exception: bool = True,
+    ) -> "Span":
+        return self._get_current_tracer().start_span(
+            name=name,
+            context=context,
+            kind=kind,
+            attributes=attributes,
+            links=links,
+            start_time=start_time,
+            record_exception=record_exception,
+            set_status_on_exception=set_status_on_exception,
+        )
+
+    @contextmanager  # type: ignore
+    def start_as_current_span(
+        self,
+        name: str,
+        context: typing.Optional[Context] = None,
+        kind: SpanKind = SpanKind.INTERNAL,
+        attributes: types.Attributes = None,
+        links: typing.Sequence[Link] = (),
+        start_time: typing.Optional[int] = None,
+        record_exception: bool = True,
+        set_status_on_exception: bool = True,
+    ) -> typing.Iterator["Span"]:
+        with self._get_current_tracer().start_as_current_span(
+            name=name,
+            context=context,
+            kind=kind,
+            attributes=attributes,
+            links=links,
+            start_time=start_time,
+            record_exception=record_exception,
+            set_status_on_exception=set_status_on_exception,
+        ) as span:
+            yield span
+
+    @contextmanager  # type: ignore
+    def use_span(
+        self, span: "Span", end_on_exit: bool = False,
+    ) -> typing.Iterator[None]:
+        with self._get_current_tracer().use_span(
+            span=span, end_on_exit=end_on_exit,
+        ) as context:
+            yield context
+
+
 _TRACER_PROVIDER = None
 
 
+@functools.lru_cache(maxsize=None)  # type: ignore
+def _get_current_tracer(*args: typing.Any, **kwargs: typing.Any) -> "Tracer":
+    tracer_provider = get_tracer_provider()
+    return tracer_provider.get_tracer(*args, **kwargs)  # type: ignore
+
+
 def get_tracer(
     instrumenting_module_name: str,
     instrumenting_library_version: str = "",
     tracer_provider: typing.Optional[TracerProvider] = None,
 ) -> "Tracer":
     """Returns a `Tracer` for use by the given instrumentation library.
 
-    This function is a convenience wrapper for
-    opentelemetry.trace.TracerProvider.get_tracer.
-
-    If tracer_provider is ommited the current configured one is used.
+    If tracer_provider is omitted it returns a _ProxyTracer
+    which redirects calls to a current instrumentation library.
     """
     if tracer_provider is None:
```

Comment:
I'm a bit confused by this. Looks like we are assuming user will always pass a valid/configured tracer_provider. Why not keep existing behavior and check if the provider passed in to the function or fetched by get_tracer_provider() is an initialized one or a default/noop one, and then return a proxy or real tracer accordingly. 

Also, doesn't this only solve the issue when users call `trace.get_tracer()`? What if someone calls `get_tracer_provider().get_tracer()` instead? Wouldn't we end up with the same problem? Unless I'm missing something, looks like we are only solving the problem partially.

Would it make more sense to patch API package's default tracer provider and have it always return a proxy tracer?

Reply:
> I'm a bit confused by this. Looks like we are assuming user will always pass a valid/configured tracer_provider. Why not keep existing behavior and check if the provider passed in to the function or fetched by get_tracer_provider() is an initialized one or a default/noop one, and then return a proxy or real tracer accordingly.

`get_tracer_provider` returns either default one, configured or set by `set_tracer_provider`. How can you distinguish them?

> Also, doesn't this only solve the issue when users call trace.get_tracer()? What if someone calls get_tracer_provider().get_tracer() instead? Wouldn't we end up with the same problem? Unless I'm missing something, looks like we are only solving the problem partially.

There was the same question above: https://github.com/open-telemetry/opentelemetry-python/pull/1445#issuecomment-779300509

> Would it make more sense to patch API package's default tracer provider and have it always return a proxy tracer?

Then you'll need another "real default tracer" that "default proxy tracer" will use when no real provided. Which problem will it solve?